### PR TITLE
Dev simulator v2

### PR DESCRIPTION
**Pull request by hypervillain**
https://linear.app/prismic/issue/SM-897/aauser-when-i-click-on-preview-slice-it-opens-the-slice-simulator-in-a

- Removed setup drawer from SliceBuilder
- Cleaned saga call to reflect new test on sm.json
- Added tests for saga rewrite
- Prepared code in simulator (checkSetup and reflect UI)
- Reopen existing Simulator tab based on tab ID (best effort)
- Open the simulator without pre-check
- Track simulator open event

Note that some stuff is left commented as I'll tackle them in another PR

### CODE REVIEW

File: packages/slice-machine/components/ScreenshotChangesModal/Views/index.tsx [hunk 1]
```diff
@@ -44,22 +38,19 @@ const VariationScreenshot: React.FC<{
...
-  }));
-  const { generateSliceCustomScreenshot, checkSimulatorSetup } =
-    useSliceMachineActions();
+  const { isLoadingScreenshot, isCheckingSimulatorSetup } = useSelector(
+    (state: SliceMachineStoreType) => ({
+      isLoadingScreenshot: isLoading(
+        state,
+        LoadingKeysEnum.GENERATE_SLICE_CUSTOM_SCREENSHOT
+      ),
+      isCheckingSimulatorSetup: isLoading(
```
Comment (danigo99) [inline packages/slice-machine/components/ScreenshotChangesModal/Views/index.tsx]: We don't need this anymore right? The button will just immediately open the simulator

File: packages/slice-machine/src/modules/simulator/index.ts [hunk 2]
```diff
@@ -308,14 +215,21 @@ function* trackOpenSetupDrawerSaga() {
 
 // Saga watchers
 function* watchCheckSetup() {
-  yield takeLatest(
+  yield takeLeading(
```
Comment (danigo99) [inline packages/slice-machine/src/modules/simulator/index.ts:201]: ❓ out of interest, why use `takeLeading` here instead of `takeLatest`

File: packages/slice-machine/src/modules/simulator/index.ts [hunk 3]
```diff
@@ -206,96 +146,63 @@ function* checkSetupSaga(
       checkSimulatorSetup
     );
 
-    // All the backend checks are ok ask for the frontend Iframe check
-    if ("ok" === setupStatus.manifest && "ok" === setupStatus.dependencies) {
+    if (setupStatus.manifest === "ok") {
```
Comment (danigo99) [inline packages/slice-machine/src/modules/simulator/index.ts:135]: Wouldn't it be easier to use a boolean here instead of the string? As far as I can tell "ko" and null and treated the same so could be condensed into `isManifestOk`

File: packages/slice-machine/src/modules/modal/index.ts [hunk 4]
```diff
@@ -34,13 +34,10 @@ export const modalReducer: Reducer<ModalStoreType, ModalActions> = (
 ) => {
   switch (action.type) {
     case getType(modalCloseCreator):
-      return {
-        ...state,
-        [action.payload.modalKey]: false,
-      };
+      return initialState;
```
Comment (danigo99) [inline packages/slice-machine/src/modules/modal/index.ts:35]: I'm not sure I agree with this, since we don't know if we'll want to close all modals together in future. 

But if we do this, we could remove all the functions in `useSliceMachineActions` for each specific closeModal and only have one `closeModals` that closes all of them

File: packages/slice-machine/src/modules/simulator/index.ts (hunk 3 above)
Comment (hypervillain) [reply packages/slice-machine/src/modules/simulator/index.ts:135]: I'm a bit unclear with this and I remember I raised the same issue when I reviewed this part of the code. But differentiating between ko and null makes things easier when you create a race condition between different events (iframeCheckOk, iframeCheckKO and timeout). It's treated the same way in the saga but it offers more flexibility in the UI. Still not super clean indeed

File: packages/slice-machine/src/modules/modal/index.ts (hunk 4 above)
Comment (hypervillain) [reply packages/slice-machine/src/modules/modal/index.ts:35]: Good point for closeModals. I checked with @Heyrows and we can go forward with this

File: packages/slice-machine/src/modules/simulator/index.ts (hunk 2 above)
Comment (hypervillain) [reply packages/slice-machine/src/modules/simulator/index.ts:201]: `takeLatest` takes (well) the latest fork and cancels any previous call. `takeLeading` prevents any subsequent fork while the saga is playing!

File: packages/slice-machine/components/ScreenshotChangesModal/Views/index.tsx (hunk 1 above)
Comment (hypervillain) [reply packages/slice-machine/components/ScreenshotChangesModal/Views/index.tsx]: true!

File: packages/slice-machine/components/Simulator/components/FailedConnect/index.tsx [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  <FullPage>
+    <Heading as="h3" sx={{ color: "textClear", mb: 2 }}>
+      Slice Machine can't render your Slice
+    </Heading>
+    <Text
+      sx={{ color: "greyIcon", mt: 1, maxWidth: "500px", textAlign: "center" }}
+    >
+      Ensure your website's development server is running by typing{" "}
+      <Text as="code" variant="styles.inlineCode">
+        npm run dev
```
Comment (MarcMcIntosh) [inline packages/slice-machine/components/Simulator/components/FailedConnect/index.tsx:34]: good advice :)

File: packages/slice-machine/components/Simulator/components/IframeRenderer/index.tsx [hunk 6]
```diff
@@ -93,7 +105,14 @@ const IframeRenderer: React.FunctionComponent<IframeRendererProps> = ({
       .catch(() => {
         connectToSimulatorFailure();
       });
-  }, [client, screenDimensions, apiContent]);
+  }, [client, screenDimensions, apiContent, simulatorUrl]);
```
Comment (MarcMcIntosh) [inline packages/slice-machine/components/Simulator/components/IframeRenderer/index.tsx:108]: are any of these objects or arrays ?

File: packages/slice-machine/lib/builders/SliceBuilder/index.tsx [hunk 7]
```diff
@@ -87,30 +61,23 @@ const SliceBuilder: React.FC<SliceBuilderProps> = ({
     }
   }, [data]);
```
Comment (MarcMcIntosh) [inline packages/slice-machine/lib/builders/SliceBuilder/index.tsx:62]: hmm, adding an object to evaluate :/

File: packages/slice-machine/package.json [hunk 8]
```diff
@@ -12,7 +12,7 @@
     "node": ">=14"
   },
   "scripts": {
-    "dev": "concurrently \"next dev\" \"npm run dev --prefix ../../e2e-projects/next\"",
```
Comment (MarcMcIntosh) [inline packages/slice-machine/package.json]: which project does this get run in now?

Comment (MarcMcIntosh) [reply packages/slice-machine/package.json]: also doesn't need to be concurrently if only one command is run

File: packages/slice-machine/components/Simulator/components/Header/index.tsx [hunk 9]
```diff
@@ -33,13 +33,15 @@ type PropTypes = {
...
   toggleIsDisplayEditor: () => void;
+  disabled: boolean;
 };
 
 const Header: React.FunctionComponent<PropTypes> = ({
   slice,
   variation,
   isDisplayEditor,
   toggleIsDisplayEditor,
+  disabled,
```
Comment (danigo99) [inline packages/slice-machine/components/Simulator/components/Header/index.tsx]: ❓ Struggling to see if this is supposed to disable the whole header. If it's only for the variations selector + switch, should we name it something more specific?

File: packages/slice-machine/components/Simulator/index.tsx [hunk 10]
```diff
@@ -81,26 +83,32 @@ const Simulator: ComponentWithSliceProps = ({ slice, variation }) => {
...
   const currentState: UiState = (() => {
     if (manifestStatus === "ok") {
-      if (isWaitingForIFrameCheck) {
-        return UiState.LOADING_IFRAME;
+      if (isWaitingForIFrameCheck || !iframeStatus) {
+        return iframeCheckFailedOnce
+          ? UiState.FAILED_CONNECT
+          : UiState.LOADING_IFRAME;
+      }
+      if (iframeStatus === "ko") {
```
Comment (danigo99) [inline packages/slice-machine/components/Simulator/index.tsx:97]: 👍 Ah this makes the difference a lot clearer!

File: packages/slice-machine/src/hooks/useCurrentSlice.tsx [hunk 11]
```diff
@@ -0,0 +1,42 @@
+import { selectCurrentSlice } from "@src/modules/selectedSlice/selectors";
+import useSliceMachineActions from "@src/modules/useSliceMachineActions";
+import { SliceMachineStoreType } from "@src/redux/type";
+
+import { useRouter } from "next/router";
```
Comment (MarcMcIntosh) [inline packages/slice-machine/src/hooks/useCurrentSlice.tsx:7]: you know the router state is in redux aswell, with operations coming from `connected-next-router`

File: packages/slice-machine/src/layouts/WithSlice.tsx [hunk 12]
```diff
@@ -0,0 +1,27 @@
+import { ComponentUI } from "@lib/models/common/ComponentUI";
+import { VariationSM } from "@slicemachine/core/build/models";
+import useCurrentSlice from "@src/hooks/useCurrentSlice";
+import { AppProps } from "next/app";
+
+import { useRouter } from "next/router";
```
Comment (MarcMcIntosh) [inline packages/slice-machine/src/layouts/WithSlice.tsx]: is `connected-next-router` viable here?

File: packages/slice-machine/src/modules/useSliceMachineActions.ts [hunk 13]
```diff
@@ -93,38 +93,25 @@ const useSliceMachineActions = () => {
...
   const openRenameCustomTypeModal = () =>
     dispatch(modalOpenCreator({ modalKey: ModalKeysEnum.RENAME_CUSTOM_TYPE }));
   const openScreenshotPreviewModal = () =>
     dispatch(modalOpenCreator({ modalKey: ModalKeysEnum.SCREENSHOT_PREVIEW }));
-  const closeScreenshotPreviewModal = () =>
-    dispatch(modalCloseCreator({ modalKey: ModalKeysEnum.SCREENSHOT_PREVIEW }));
   const openSimulatorSetupModal = () =>
     dispatch(modalOpenCreator({ modalKey: ModalKeysEnum.SIMULATOR_SETUP }));
-  const closeSimulatorSetupModal = () =>
-    dispatch(modalCloseCreator({ modalKey: ModalKeysEnum.SIMULATOR_SETUP }));
```
Comment (danigo99) [inline packages/slice-machine/src/modules/useSliceMachineActions.ts]: 👨‍🍳 👍

File: packages/slice-machine/src/theme.ts [hunk 14]
```diff
@@ -765,8 +768,8 @@ const AppTheme = (): Theme =>
       },
       inlineCode: {
         fontFamily: "monospace",
-        color: "error",
-        bg: "muted",
+        color: "#6E56CF",
+        bg: "#F1EEFE",
```
Comment (danigo99) [inline packages/slice-machine/src/theme.ts:776]: Why this change? Would be good to avoid string literal colours outside the colours section at the top

File: packages/slice-machine/src/modules/availableCustomTypes/index.ts [hunk 15]
```diff
@@ -177,9 +177,7 @@ export function* createCustomTypeSaga({
     );
     yield call(saveCustomType, newCustomType, {});
     yield put(createCustomTypeCreator.success({ newCustomType }));
-    yield put(
-      modalCloseCreator({ modalKey: ModalKeysEnum.CREATE_CUSTOM_TYPE })
```
Comment (MarcMcIntosh) [inline packages/slice-machine/src/modules/availableCustomTypes/index.ts:181]: linter warning on line 11

File: packages/slice-machine/src/theme.ts [hunk 16]
```diff
@@ -68,7 +68,7 @@ const AppTheme = (): Theme =>
       warning: "#E67E22",
       warning02: "#ED811C",
       sidebar: "#F1F1F4",
-      link: "#5163BA",
+      link: "#6E56CF",
```
Comment (danigo99) [inline packages/slice-machine/src/theme.ts:75]: ❓ This doesn't cause side-effects elsewhere?

File: packages/slice-machine/components/Simulator/components/FullPage/index.tsx [hunk 17]
```diff
@@ -0,0 +1,21 @@
+import { Flex } from "theme-ui";
+
+const FullPage: React.FC = ({ children }) => (
+  <Flex
+    sx={{
+      bg: "grey01",
+      top: "128px",
+      position: "absolute",
+      height: "calc(100vh - 128px)",
```
Comment (danigo99) [inline packages/slice-machine/components/Simulator/components/FullPage/index.tsx]: ❓ Is 128px because of the header? Could the same not be accomplished only with flex to avoid the hardcoded values?

File: packages/slice-machine/src/modules/modal/index.ts [hunk 18]
```diff
@@ -34,13 +32,10 @@ export const modalReducer: Reducer<ModalStoreType, ModalActions> = (
...
     case getType(modalOpenCreator):
       return {
-        ...state,
+        ...initialState,
```
Comment (MarcMcIntosh) [inline packages/slice-machine/src/modules/modal/index.ts:38]: why not state?

File: packages/slice-machine/src/modules/simulator/index.ts [hunk 19]
```diff
@@ -6,13 +6,13 @@ import {
...
 } from "typesafe-actions";
-import { SimulatorStoreType, SetupStatus } from "./types";
+import { SimulatorStoreType } from "./types";
 import {
   call,
   fork,
   put,
   select,
-  takeLatest,
+  takeLeading,
```
Comment (MarcMcIntosh) [inline packages/slice-machine/src/modules/simulator/index.ts:10]: is there an advantage over takeLatest vs takeLeading?

File: packages/slice-machine/src/modules/simulator/index.ts [hunk 20]
```diff
@@ -23,52 +23,35 @@ import {
...
   "SIMULATOR/CHECK_SETUP.SUCCESS",
   "SIMULATOR/CHECK_SETUP.FAILURE"
 )<
   {
-    withFirstVisitCheck: boolean;
     callback?: () => void;
   },
   {
-    setupStatus: SetupStatus;
+    setupStatus: SimulatorStoreType["setupStatus"];
```
Comment (MarcMcIntosh) [inline packages/slice-machine/src/modules/simulator/index.ts:49]: would `Pick<SimulatorStoreType, "setupStatus">` also work?

File: packages/slice-machine/src/modules/simulator/index.ts [hunk 21]
```diff
@@ -79,68 +62,33 @@ export const connectToSimulatorIframeCreator = createAsyncAction(
...
 
 // Selectors
-export const selectIsSetupDrawerOpen = (
+export const selectSetupStatus = (
   state: SliceMachineStoreType
-): boolean => state.simulator.setupDrawer.isOpen;
-
-export const selectSetupStatus = (state: SliceMachineStoreType): SetupStatus =>
-  state.simulator.setupStatus;
+): SimulatorStoreType["setupStatus"] => state.simulator.setupStatus;
```
Comment (MarcMcIntosh) [inline packages/slice-machine/src/modules/simulator/index.ts:71]: ca use `Pick<SimulatorStoreType, "setupStatus">` here if you want

File: packages/slice-machine/src/modules/simulator/index.ts [hunk 22]
```diff
@@ -206,96 +137,65 @@ function* checkSetupSaga(
...
   } catch (error) {
     // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
     yield put(checkSimulatorSetupCreator.failure(error as Error));
   }
 }
 
+function* connectToSimulatorIframe() {
+  yield put(connectToSimulatorIframeCreator.request());
+  const {
+    timeout,
```
Comment (MarcMcIntosh) [inline packages/slice-machine/src/modules/simulator/index.ts:159]: linting issue with `any`

File: packages/slice-machine/src/modules/simulator/index.ts [hunk 23]
```diff
@@ -206,96 +137,65 @@ function* checkSetupSaga(
...
+  }
+  yield put(connectToSimulatorIframeCreator.success());
+}
+
 export function* failCheckSetupSaga() {
   const framework = (yield select(getFramework)) as ReturnType<
     typeof getFramework
   >;
+  // This should be used in track event
+  framework;
```
Comment (MarcMcIntosh) [inline packages/slice-machine/src/modules/simulator/index.ts]: why is it just hanging there?

File: packages/slice-machine/src/modules/simulator/index.ts (hunk 2 above)
Comment (MarcMcIntosh) [reply packages/slice-machine/src/modules/simulator/index.ts:201]: interesting

File: packages/slice-machine/tests/src/modules/simulator.test.ts [hunk 24]
```diff
@@ -63,122 +65,76 @@ describe("[Simulator module]", () => {
...
-    it("should open the fifth step on nuxt", () => {
+  describe("[failCheckSetupSaga]", () => {
+    it("should early return if the framework doesn't support the simulator feature", () => {
       const saga = testSaga(failCheckSetupSaga);
 
       saga.next().select(getFramework);
-      saga.next("nuxt").select(selectIsSimulatorAvailableForFramework);
-      saga.next(true).put(openSetupDrawerCreator({ stepToOpen: 5 }));
-      saga.next().isDone();
+      saga.next("next").select(selectIsSimulatorAvailableForFramework);
```
Comment (MarcMcIntosh) [inline packages/slice-machine/tests/src/modules/simulator.test.ts]: it is though isn't it?

File: packages/slice-machine/tests/src/modules/simulator.test.ts [hunk 25]
```diff
@@ -63,122 +65,76 @@ describe("[Simulator module]", () => {
...
-          openedStep: 0,
-          isOpen: true,
-        },
-      };
-
-      const action = closeSetupDrawerCreator();
-
-      expect(simulatorReducer(initialState, action)).toEqual(
-        dummySimulatorState
+      const saga = testSaga(checkSetupSaga, setupSimulatorAction);
```
Comment (MarcMcIntosh) [inline packages/slice-machine/tests/src/modules/simulator.test.ts:94]: fyi: expect saga give you more coverage ;)

File: packages/slice-machine/tests/server/apiController/simulator.test.ts [hunk 26]
```diff
@@ -57,130 +55,5 @@ describe("simulator controller", () => {
       requestWithPreviewUrl
     );
     expect(previewCheckResponse.manifest).toBe("ok");
-    expect(previewCheckResponse.dependencies).toBe("ko");
-  });
-
-  test("it should return ok when all the legacy next required deps are installed", async () => {
```
Comment (MarcMcIntosh) [inline packages/slice-machine/tests/server/apiController/simulator.test.ts:63]: do we no longer check that the user has installed all the deps?

File: packages/slice-machine/src/hooks/useCurrentSlice.tsx [hunk 27]
```diff
@@ -0,0 +1,42 @@
...
+import { useEffect } from "react";
+import { useSelector } from "react-redux";
+
+function useCurrentSlice() {
```
Comment (MarcMcIntosh) [inline packages/slice-machine/src/hooks/useCurrentSlice.tsx]: should this be typed?

File: packages/slice-machine/lib/builders/SliceBuilder/SideBar/index.tsx [hunk 28]
```diff
@@ -35,23 +34,18 @@ const SideBar: React.FunctionComponent<SideBarProps> = ({
   component,
   variation,
 }) => {
+  const router = useRouter();
```
Comment (MarcMcIntosh) [inline packages/slice-machine/lib/builders/SliceBuilder/SideBar/index.tsx]: router should already be in the redux store :)

**Review by MarcMcIntosh — COMMENTED** [review]
Left a few comments :)

Shat Im wondering is if  unit test for slimutalor.tsx, need to br written, what would be all the external deps and expected behavours?

File: packages/slice-machine/lib/builders/SliceBuilder/index.tsx (hunk 7 above)
Comment (hypervillain) [reply packages/slice-machine/lib/builders/SliceBuilder/index.tsx:62]: at some point I guess? But it's not for today!

File: packages/slice-machine/package.json (hunk 8 above)
Comment (hypervillain) [reply packages/slice-machine/package.json]: I re-added it in another PR that will be merged in this one! Mistake sorry

File: packages/slice-machine/src/hooks/useCurrentSlice.tsx (hunk 11 above)
Comment (hypervillain) [reply packages/slice-machine/src/hooks/useCurrentSlice.tsx:7]: No I didn't know! Should I always use redux instead?

File: packages/slice-machine/components/Simulator/components/Header/index.tsx (hunk 9 above)
Comment (hypervillain) [reply packages/slice-machine/components/Simulator/components/Header/index.tsx]: Sort of, it also disables the switch and will disable the save mocks button. So it sort of disables the whole Header but I can change that if you have something more specific in mind. `actionsDisabled`?

File: packages/slice-machine/src/theme.ts (hunk 14 above)
Comment (hypervillain) [reply packages/slice-machine/src/theme.ts:776]: I agree but the whole theme needs to be refactored, and I can't say for sure yet what belongs in colors and what does not

File: packages/slice-machine/components/Simulator/components/FullPage/index.tsx (hunk 17 above)
Comment (hypervillain) [reply packages/slice-machine/components/Simulator/components/FullPage/index.tsx]: I'm not sure because position is absolute here

File: packages/slice-machine/src/modules/modal/index.ts (hunk 18 above)
Comment (hypervillain) [reply packages/slice-machine/src/modules/modal/index.ts:38]: Because it resets everything correctly

File: packages/slice-machine/src/modules/simulator/index.ts (hunk 22 above)
Comment (hypervillain) [reply packages/slice-machine/src/modules/simulator/index.ts:159]: I'm unable to fix it, would you mind taking a look?

File: packages/slice-machine/tests/server/apiController/simulator.test.ts (hunk 26 above)
Comment (hypervillain) [reply packages/slice-machine/tests/server/apiController/simulator.test.ts:63]: nope!

File: packages/slice-machine/components/Simulator/components/IframeRenderer/index.tsx (hunk 6 above)
Comment (hypervillain) [reply packages/slice-machine/components/Simulator/components/IframeRenderer/index.tsx:108]: yes

File: packages/slice-machine/src/theme.ts (hunk 16 above)
Comment (hypervillain) [reply packages/slice-machine/src/theme.ts:75]: It's actually used nowhere but in my component :/

File: packages/slice-machine/src/modules/modal/index.ts (hunk 18 above)
Comment (MarcMcIntosh) [reply packages/slice-machine/src/modules/modal/index.ts:38]: okay

File: packages/slice-machine/components/Simulator/components/FullPage/index.tsx (hunk 17 above)
Comment (danigo99) [reply packages/slice-machine/components/Simulator/components/FullPage/index.tsx]: Does it need to be absolute?

File: packages/slice-machine/pages/_app.tsx [hunk 29]
```diff
@@ -22,7 +22,7 @@ import "src/css/drawer.css";
 import "src/css/toaster.css";
 import "src/css/intercom.css";
 
-import "highlight.js/styles/atom-one-dark.css";
+import "src/css/hljs.css";
```
Comment (MarcMcIntosh) [inline packages/slice-machine/pages/_app.tsx:25]: are these the same?

**Comment by arnaudlewis** [comment]
merged in #739
